### PR TITLE
[git] Implement performance hack for MacOS from Magit Manual

### PR DESCRIPTION
**Pull request by Kazark**
I indeed observed the slowness and this helped. It may not be necessary for every old version of Mac OS but should be harmless on them all.

### CODE REVIEW

File: layers/+source-control/git/packages.el [hunk 1]
```diff
@@ -234,6 +234,9 @@
       ;; whitespace
       (define-key magit-status-mode-map (kbd "C-S-w")
         'spacemacs/magit-toggle-whitespace)
+      ;; https://magit.vc/manual/magit/MacOS-Performance.html
+      (when (eq system-type 'darwin)
+        (setq magit-git-executable (s-trim (shell-command-to-string "which git"))))
```
Comment (tko) [inline layers/+source-control/git/packages.el:239]: I'm probably nitpicking but going through shell raises an eyebrow, and in the odd edge case of not having git setting the executable to empty string(?) feels slightly wrong. I would've thought something like this, but I'm not sure if deferring to shell would actually be preferrable
```elisp
(when-let ((git (executable-find "git")))
  (setq magit-git-executable git))
```
I suppose one could argue this should be safe on all OSs and there might be some benefit for being consistent, but :shrug:

**Comment by Kazark** [comment]
Thanks @tko. I did leave the `(when (eq system-type 'darwin) ...)` as a marker for the reason why the code exists at all.

File: layers/+source-control/git/packages.el (hunk 1 above)
Comment (smile13241324) [reply layers/+source-control/git/packages.el]: I tend to agree to @tko, this code should be fairly save for all OSes so for the sake of consistency please remove the `only darwin when` please.

Comment (Kazark) [reply layers/+source-control/git/packages.el]: @tko @smile13241324 alright, removed the condition in ad941de

**Comment by Kazark** [comment]
The "Validate PR" build is broken. I wonder why.

**Review by smile13241324 — APPROVED** [review]
Looks good now.

**Comment by smile13241324** [comment]
Looks like you have merged develop onto your branch which should have been a rebase, that's why the PR check fails. I will squash your commits anyway so it does not hurt.

**Comment by smile13241324** [comment]
Thank you for contributing to spacemacs 💜, I have cherry picked your commit into develop.

**Comment by Kazark** [comment]
Ah, rebase, don't merge. Got it---thanks.

**Comment by bcc32** [comment]
FYI, this PR breaks the use of magit over TRAMP when the path to git is different on the local and remote machines.  For example, on my mac, magit-git-executable gets set to `/usr/local/bin/git` (installed via homebrew), but this breaks when I try to access files remotely on my NixOS server using TRAMP, where git is at `/run/current-system/sw/bin/git`.

Also, as it happens, on macOS Big Sur (11.1), I see no speed difference in `magit-status` with and without this patch.  I think, at the very least, this code should be re-guarded behind an OS check.  IMO, it seems weird to lock magit into using whatever git was found in path at startup.

**Comment by Kazark** [comment]
@bcc32 Hm, maybe I should just revert this PR entirely. I didn't understand that there were conditions under which it could be undesirable.

**Comment by bcc32** [comment]
@Kazark Since you observed the slowness without this patch, I would be sad to lose it entirely.  Maybe it could be opt-in behind a configuration option, or simply a suggestion in the git layer docs?

**Comment by Kazark** [comment]
@bcc32 is there a way to detect the presence of TRAMP, and disable it in that case?

**Comment by bcc32** [comment]
The definition of `magit-git-executable` makes me think there's no current way to do this:

```
(defcustom magit-git-executable
  ;; Git might be installed in a different location on a remote, so
  ;; it is better not to use the full path to the executable, except
  ;; on Window where we would otherwise end up using one of the
  ;; wrappers "cmd/git.exe" or "cmd/git.cmd", which are much slower
  ;; than using "bin/git.exe" directly.
```

This suggests to me that, on Windows, magit across TRAMP just doesn't work by default due to lack of magit support for changing `magit-git-executable` dynamically. :(

**Comment by Kazark** [comment]
@bcc32 I meant: detect if TRAMP is installed/used in that Emacs at the time that I set `magit-git-executable`, and just don't set it. I don't use TRAMP, and I don't know much about it.

**Comment by bcc32** [comment]
Ah, my bad.  Since TRAMP is included with Emacs, it's always installed.

**Comment by Kazark** [comment]
What if I reguarded the code, with a specific check for Mac OS version? Based on your results, perhaps whatever the root problem here is fixed in Big Sur? (and I wonder how I would determine that...)

**Comment by bcc32** [comment]
That sounds fine to me.  I'm not really sure which versions it would be appropriate to enable the patch for.  The PR that originally introduced this suggestion in the magit docs is magit/magit#4069, which links to some reddit threads with discussions.

**Comment by lebensterben** [comment]
@Kazark 
In my opinion this should be reported to upstream.
First, the optimisation trick comes from magit's manual, so its impact goes beyond spacemacs users.
Second, if macOS should be treated specially (as opposed to other OS), it should be done so in magit itself.

**Comment by Kazark** [comment]
@lebensterben excellent point. Will raise it on the issue that @bcc32 linked to.

**Comment by jjgriego** [comment]
Just gonna leave a breadcrumb for anyone else who gets here and needs a fix, since it was a little nonobvious, sorry for resurrecting a zombie PR. 

I fixed the TRAMP issue for myself locally by adding to `spacemacs/user-config`:

`(eval-after-load 'magit '(setq magit-git-executable "git"))`

I can understand why we might want to leave this in until magit can patch since an error ("where's your git?!") is at least a little bit more to work with when debugging than mysterious slowness.

**Comment by bcc32** [comment]
> I can understand why we might want to leave this in until magit can patch since an error ("where's your git?!") is at least a little bit more to work with when debugging than mysterious slowness.

I agree in principle, although I didn't actually experience an error.  Instead, when I pressed `SPC g s` in a repository over TRAMP, I was prompted for the git repo root.  (But I'm already in a git repo!).  So it is confusing and nonobvious either way.

**Pull request merged**